src/Features/Core/Portable/InlineMethod/AbstractInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,382 @@
+                // int Callee() { throw new Exception();}
+                // Note here throw statement is changed to throw expression after inlining
+                // If this is a throw expression, the check is the same
+                // 1. If it is invoked as ExpressionStatement
+                // Example:
+                // Before:
+                // void Caller() { Callee(); }
+                // void Callee() => throw new Exception();
+                // After:
+                // void Caller() { throw new Exception(); }
+                // void Callee() => throw new Exception();
+                // Note here throw expression is converted to throw statement
+                // 2. If it is invoked in a place allow throw expression
+                // Example:
+                // Before:
+                // void Caller(bool flag) { var x = flag ? Callee() : 1; }
+                // int Callee() => throw new Exception();
+                // After:
+                // void Caller() { var x = flag ? throw new Exception() : 1; }
+                // int Callee() => throw new Exception();
+                if (!CanBeReplacedByThrowExpression(calleeMethodInvocationNode)
+                    && !_syntaxFacts.IsExpressionStatement(calleeMethodInvocationNode.Parent))
+                {
+                    return;
+                }
+            }
+
+            var invocationOperation = semanticModel.GetOperation(calleeMethodInvocationNode, cancellationToken) as IInvocationOperation;
+            if (invocationOperation == null)
+            {
+                return;
+            }
+
+            var codeActions = GenerateCodeActions(
+                document,
+                calleeMethodInvocationNode,
+                calleeMethodSymbol,
+                calleeMethodNode,
+                inlineExpression,
+                statementContainsCallee,
+                invocationOperation);
+
+            var nestedCodeAction = new CodeAction.CodeActionWithNestedActions(
+                string.Format(FeaturesResources.Inline_0, calleeMethodSymbol.ToNameDisplayString()),
+                codeActions,
+                isInlinable: true);
+
+            context.RegisterRefactoring(nestedCodeAction, calleeMethodInvocationNode.Span);
+        }
+
+        private ImmutableArray<CodeAction> GenerateCodeActions(
+            Document document,
+            TInvocationSyntax calleeMethodInvocationNode,
+            IMethodSymbol calleeMethodSymbol,
+            TMethodDeclarationSyntax calleeMethodNode,
+            TExpressionSyntax inlineExpression,
+            TStatementSyntax statementContainsCallee,
+            IInvocationOperation invocationOperation)
+        {
+            var calleeMethodName = calleeMethodSymbol.ToNameDisplayString();
+            var codeActionKeepsCallee = new MySolutionChangeAction(
+                string.Format(FeaturesResources.Inline_0_and_keep_1, calleeMethodName, calleeMethodName),
```

Comment:
`Inline and keep '{0}'`.  There no need to say `Inline 'Foo' and keep 'Foo'` when we can say `Inline and keep 'Foo'`.